safesocket/safesocket_darwin.go

```diff
@@ -17,23 +20,246 @@ import (
 	"sync"
 	"time"
 
+	"golang.org/x/sys/unix"
 	"tailscale.com/version"
 )
 
 func init() {
 	localTCPPortAndToken = localTCPPortAndTokenDarwin
 }
 
-// localTCPPortAndTokenMacsys returns the localhost TCP port number and auth token
+// SameUserProofTokenLength is the length in bytes of the token used to
+// authenticate safesocket connections
+const (
+	sameUserProofTokenLength = 10
+	SyxExtLibraryDir         = "/Library/Tailscale"
+)
+
+type darwinCredentials struct {
+	mu    sync.Mutex
+	creds struct {
+		token string
+		port  int
+	}
+}
+
+var (
+	// explicit credentials set by SetCredentials
+	dc darwinCredentials
+	// file descriptor to the open sameuserproof file for mac App Store
+	sameuserproofFD *os.File
+
+	// For testing only to force macsys
+	isMacSysExt = version.IsMacSysExt()
+	// For testing only to skip the connection verification
+	verifyConn = true
+)
+
+// There are two ways this binary can be run: as the Mac App Store or Standalone binary,
+// or a separate CLI binary that somebody built or download.
+//
+// The  macOS binaries can communicate directly via XPC with the tailscaled process
+// and must always call SetCredentials directly.  If we are a separate CLI binary, we must
+// source the credentials from the sameuserproof file which we will write down in
+// the supplied sharedDir.
+
+// LocalTCPPortAndTokenDarwin returns the localhost TCP port number and auth token
+// either generated, or sourced from the tailscaled process.
+func localTCPPortAndTokenDarwin() (port int, token string, err error) {
+	dc.mu.Lock()
+	defer dc.mu.Unlock()
+
+	if dc.creds.port != 0 && dc.creds.token != "" {
+		return dc.creds.port, dc.creds.token, nil
+	}
+
+	// As a fallback, we'll query the sameuserproof file.
+	port, token, err = PortAndTokenFromSameUserProof()
+	return port, token, err
+}
+
+// SetCredentials sets and explicit token and port used to authenticate safesocket
```

Comment:
s/and/an/

and trailing period

But this is only for macOS? Or also iOS? say that?